### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by gabi-250**
This PR contains a simple benchmark for the interpreter.

### CODE REVIEW

File: benchmark.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+if [ ! -f ./interpreter-c ] && [ $# -lt 1 ]; then
```
Comment (vext01) [inline benchmark.sh]: I find this check slightly odd. If the interpreter is not built I will see an message about needing an argument.

I would just assume the interpreter is built to be honest.

File: benchmark.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    echo "[ERROR] ./interpreter-c not found: Need an argument. Usage: ./benchmark.sh [PYTHONPATH]"
+else
+    if [ ! -f ./interpreter-c ]; then
+        $1 -O0 interpreter.py
+    fi
+    make clean
+    make
+    ./multitime -n5 ./interpreter-c fibonacci.bc
```
Comment (vext01) [inline benchmark.sh]: On second thoughts, if you put the multitime part into a Make target, make can check the interpreter is built using dependencies.

File: fibonacci.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        n = n * 10 + *(number + i) - 48;
+        ++i;
+    }
+    return n;
+}
+
+int fibonacci_rec(int n)
+{
+    if(n < 2) {
+        return 1;
```
Comment (vext01) [inline fibonacci.c]: print an error?

File: fibonacci.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    else {
+        return fibonacci_rec(n - 1) + fibonacci_rec(n - 2);
```
Comment (vext01) [inline fibonacci.c]: Does it make sense to have both the iterative and recursive versions in the same file? How much code do they share?

**Comment by vext01** [comment]
A few comments.

File: fibonacci.c (hunk 3 above)
Comment (gabi-250) [inline fibonacci.c]: Why should I print an error? It's simply the base case for recursion.

Comment (vext01) [inline fibonacci.c]: Sorry, I misread. Carry on.

File: fibonacci.c (hunk 4 above)
Comment (gabi-250) [inline fibonacci.c]: I'll place them in separate files.

File: benchmark.sh (hunk 2 above)
Comment (gabi-250) [inline benchmark.sh]: I could try that. Should I keep adding stuff to the Makefile I already have?

Comment (vext01) [inline benchmark.sh]: Sure, why not.

**Comment by gabi-250** [comment]
I've slightly changed how the benchmarks can be used. I added multitime as a target as you've suggested. Now you have to pass `PYTHONPATH` as a script argument.

File: Makefile [hunk 5]
```diff
@@ -5,10 +5,22 @@ LDFLAGS := $(shell llvm-config --cxxflags --ldflags --libs core native --system-
...
+fibonacci_iter.bc: fibonacci_iter.c
+	$(CLANG_CC) -emit-llvm fibonacci_iter.c -c -O2  -o fibonacci_iter.bc
+
+fibonacci_rec.bc: fibonacci_rec.c
+	$(CLANG_CC) -emit-llvm fibonacci_rec.c -c -O2  -o fibonacci_rec.bc
+
+interpreter-c: interpreter.py
+	$(PYTHONPATH) -O0 interpreter.py
+
+multitime: ./interpreter-c
```
Comment (vext01) [inline Makefile]: One more line for the iterative version.

File: Makefile [hunk 6]
```diff
@@ -5,10 +5,22 @@ LDFLAGS := $(shell llvm-config --cxxflags --ldflags --libs core native --system-
 
 .PHONY: all clean
 
-all: test.bc
+all: test.bc fibonacci_rec.bc fibonacci_iter.bc multitime
```
Comment (vext01) [inline Makefile]: We probably dont want the benchmark running in "all" do we?

**Comment by vext01** [comment]
Two tiny comments and we are done I think.

File: Makefile (hunk 6 above)
Comment (gabi-250) [inline Makefile]: I'll do `make multitime` then and add the C files as dependencies for multitime.

**Comment by gabi-250** [comment]
Hopefully, it looks better now.

File: benchmark.sh [hunk 7]
```diff
@@ -2,5 +2,5 @@ if [ $# -ne 1 ]; then
     echo "[ERROR] Need an argument. Usage: ./benchmark.sh PYTHONPATH"
 else
     make clean
-    PYTHONPATH=$1 make
+    PYTHONPATH=$1 make multitime
```
Comment (vext01) [inline benchmark.sh]: Ah, sorry. I think we dont need this script any more, right? We just run `make multitime` (maybe rename the target to `benchmark`).

Comment (gabi-250) [inline benchmark.sh]: That's a good point. Will fix.

**Comment by gabi-250** [comment]
Okay, now you only need to set the path and have make execute the `benchmark` target.

**Comment by vext01** [comment]
Looks good. Let's squash.

**Comment by gabi-250** [comment]
Done.

**Pull request merged**